Nodal normals improvement

andrsd opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

This enables multiple normals per node.

Since the normals now come from UserObject-derived class, changes were needed in `Penetration{Locator|Thread}` and minor syntax changes to enable the geometrical system to pick the normals from the user objects.

As a side effect (positive one), this removes the `DualRestrictable` restriction from the nodal normals system.

andrsd committed:
Improving nodal normals system

1. Allow multiple normals at a single node
   If we had a square mesh and set nodal normals on the top surface and
   left surface, then there should be 2 normals at the top left corner.
   This was not possible with the previous implementation

2. Changing syntax to:
   [NodalNormals]
     [./nn_name_1]
       (the old syntax)
     [../]
     [./nn_name_2]
       (the old_syntax)
     [../]
   []
   The updated systems pick up the normals via nn_name_XYZ.
   Any system can get the normals via:
     _uo = getUserObject<NodalNormalsUserObject)(name);
     Point normal = _uo.getNormal(node_id);

3. Removing DualRestrictable restriction.

Closes #10551

andrsd committed:
Adding missing implementation for Assembly::getFENeighbor

andrsd committed:
Removing inactive code

The c-tor sets _normal_smoothing_method to NSM_EDGE_BASED. It is not
changed anywhere, so the if-statement is always false, so it does not
have to be there.

andrsd committed:
GeometricalSearchData is always created with FEProblem

We always pass FEProblemBase into GeometricalSearchData, so it keeps the
reference to FEProblemBase rather then SubProblem.

andrsd committed:
Modifying the usage of nodal normals inside geometrical search

1. Improving setters for nodal normals smoothing
2. Code follows formatting convections
3. Updating tests (ignoring nodal_normal_ variables that are leftovers
   from the previous implementation, but might be useful in future)

andrsd commented:
@bwspenc This includes some changes you should be aware of.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/147202/) on 22445c0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10556/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10556/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 76af3257c30ff9c78bc55d4f373b4eda4ad0b679`

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/147245/) on 5051c3b wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10556/site)

This comment will be updated on new commits.

andrsd committed:
Adding doco

moosebuild commented:
Job [App tests](https://www.moosebuild.org/job/152158/) on c3cf6d6 : invalidated by @andrsd

bighorn and pronghorn are patched

moosebuild commented:
Job [Private App tests](https://www.moosebuild.org/job/152159/) on c3cf6d6 : invalidated by @andrsd

blue_crab is patched

andrsd commented:
* Patch for bighorn is ready.
* valgrind failures are unrelated to this PR.
* failures in heavy module test should be fixed by #10671

permcody commented on framework/src/userobject/NodalNormalsUserObject.C:
```diff
@@ -0,0 +1,181 @@
+void
+NodalNormalsUserObject::add(const Node * node, RealGradient grad) const
+{
+  Threads::spin_mutex::scoped_lock lock(_add_contr);
```
Do we really need the lock here? Seems like you don't need it until the other `add()` method.

andrsd replied:
This is called from `NodalNormalsBoundaryNodes`. The other `add` is also called from `deserialization`. So may be just move this lock inside the if-statement below.

andrsd replied:
I thought about this more and I do not think we want to move this. This is called from different threads and if we moved the lock, we would have unprotected access into `_mesh`.

permcody commented on framework/src/userobject/NodalNormalsUserObject.C:
```diff
@@ -0,0 +1,181 @@
+  if (hasBoundary(node_boundary_ids, BoundaryRestrictable::ANY))
+  {
+    const dof_id_type & node_id = node->id();
+    Point normal_contribution(grad(0), grad(1), grad(2));
```
OK, but does this work?
`Point normal_contribution(grad);`

permcody commented on framework/src/userobject/NodalNormalsUserObject.C:
```diff
@@ -0,0 +1,181 @@
+  for (auto & it : _nodal_normals)
+  {
+    Point n = it.second;
+    Real len = std::sqrt((n(0) * n(0)) + (n(1) * n(1)) + (n(2) * n(2)));
```
https://libmesh.github.io/doxygen/classlibMesh_1_1TypeVector.html#a08025c28484d156f197faa86ceced99f

permcody commented on framework/src/userobject/NodalNormalsUserObject.C:
```diff
@@ -0,0 +1,181 @@
+    if (std::abs(len) >= 1e-13)
+    {
+      n /= len;
+      it.second = n;
```
or use `Point::unit()`

permcody commented on framework/src/userobject/NodalNormalsUserObject.C:
```diff
@@ -0,0 +1,181 @@
+
+  recv_buffers.reserve(_app.n_processors());
+  serialize(send_buffers[0]);
+  comm().allgather_packed_range((void *)(nullptr),
```
Not super scalable in terms of memory, but should be fine for awhile.

andrsd replied:
This was a bitch to get right. The major problem is in the penetration system. Think of this case: 2 disconnected bodies each on a separate CPU. Nodal normals are computed on the right body. Then, the left body projects a node on the right body and needs the normals. So, I need to do this communication so that every CPU has all the normals. Luckily, we store normals only where they were requested (the old system has them in every node of the mesh).

andrsd replied:
Also, for the record. We compute the normals typically just once per simulation. When adaptivity is on, then we need to recompute because the mesh changed. People can request computing the normals more often, but it is not typically needed.

permcody commented on modules/heat_conduction/src/dirackernels/GapHeatPointSourceMaster.C:
```diff
@@ -47,12 +50,27 @@ GapHeatPointSourceMaster::GapHeatPointSourceMaster(const InputParameters & param
+      _penetration_locator.setNodalNormalSmoothingMethod(getParam<UserObjectName>("nodal_normals"));
+    }
+    else
+      mooseError(name(), ": Unknown smoothing method.");
```
This same error checking code is in at least 3 places in this PR. Maybe factor out into a single method?

andrsd replied:
Yes, I hate this too. Even the original code had the same flaw. I will refactor that in a separate commit

permcody reviewed (changes requested):
This looks pretty good. I'm sure there's an integration problem, but it would also be nice to add doco now while it's fresh in your head.

andrsd commented on framework/include/geomsearch/PenetrationLocator.h:
```diff
@@ -83,22 +81,36 @@ class PenetrationLocator : Restartable
+  /// Tangential distance a node can be from a face and still be in contact
+  Real _tangential_tolerance;
+  /// Should we do contact normal smoothing?
+  bool _do_normal_smoothing;
```
Notice I only fixed the style of comments, but I can rearrange the member variables

andrsd replied:
So, why are the interweaved bools a problem in `PenetrationLocator`? We do not have a lot of them usually. I could see this would be a problem in `PenetrationInfo`...

permcody replied:
Not a problem, just a note.

andrsd commented on docs/content/documentation/systems/AuxKernels/framework/NodalNormalAux.md:
```diff
@@ -0,0 +1,10 @@
+<!-- MOOSE Documentation Stub: Remove this when content is added. -->
+
+# NodalNormalAux
```
Is there some guideline for the MD files?

permcody replied:
Natural writing or prose as opposed to terse documents. Anything is better than nothing. Ideally you'd take the time to put diagrams and picture up there but that does take a lot of time. Just a free flowing paragraph or two would be helpful.

andrsd replied:
I guess "this stupid class is for viz only" is too terse... But, that's what it is :-)

sapitts replied:
@andrsd I wrote this [set](http://mooseframework.org/docs/moose_docs/site/utilities/moose_docs/docs_standards/index.html) of standards for the documentation as a starting place. Please add to the standards page as you see fit.  There's also a template file in the templates/standards directory

andrsd replied:
@sapitts Thanks. I saw that, but I thought it was just for the tensor mechanics module...

andrsd commented on framework/src/actions/AddNodalNormalsAction.C:
```diff
@@ -20,82 +20,73 @@ validParams<AddNodalNormalsAction>()
-  params.addParam<std::vector<BoundaryName>>(
-      "boundary", everywhere, "The boundary ID or name where the normals will be computed");
+  params.addParam<std::vector<SubdomainName>>(
+      "block", "The list of block ids (SubdomainID) that this object will be applied");
```
I thought `DualRestrictable` was posing some integrity checking so we wanted to get rid of it and this was the only instance that required `DualRestrictable`.

permcody replied:
Yes, you are right.

andrsd commented on framework/src/actions/AddNodalNormalsAction.C:
```diff
@@ -20,82 +20,73 @@ validParams<AddNodalNormalsAction>()
-      "boundary", everywhere, "The boundary ID or name where the normals will be computed");
+  params.addParam<std::vector<SubdomainName>>(
+      "block", "The list of block ids (SubdomainID) that this object will be applied");
+  params.addRequiredParam<std::vector<BoundaryName>>(
```
Normals live on boundaries, so this has to be required. Also the syntax changed to allow multiple block with normals. Not sure, if that will screw with your test

permcody replied:
Yes - your syntax is improved. I was just taking advantage of the fact that it was an Action that was associated with three tasks (which was the error that I fixed). I'll see if there's another Action/task set that can be used in my test.

andrsd commented on framework/src/auxkernels/PenetrationAux.C:
```diff
@@ -34,8 +34,11 @@ validParams<PenetrationAux>()
-                               "Method to use to smooth normals (edge_based|nodal_normal_based)");
+  MooseEnum smoothing_method("edge_based nodal_normal_based");
+  params.addParam<MooseEnum>(
+      "normal_smoothing_method", smoothing_method, "Method to use to smooth normals");
```
This is just switching from `std::string` to `MooseEnum`. I am not implementing this - it is already in the framework - this was done by @bwspenc IIRC, I do not know what else would you like to see

permcody replied:
What is the difference between "edge-based, versus normal based"?

andrsd replied:
I think edge based is the usual stuff that comes from the sides and nodal is normals coming from this system.

andrsd commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -1299,14 +1301,15 @@ PenetrationThread::smoothNormal(PenetrationInfo * info, std::vector<PenetrationI
+      const Real len(normal.norm());
       if (len > 0)
-        info->_normal /= len;
+        normal /= len;
```
The problem with `Point::unit()` is that if the normal is zero (which can probably happen in some weird cases), the code will trip on a libmesh assert. And the original authors wanted the code to continue and deal with this situation. I am merely switching from the original approach of having normals in variables to provide them via an user object.

andrsd commented on framework/src/auxkernels/NodalNormalAux.C:
```diff
@@ -0,0 +1,49 @@
+  else if (_component_enum == "y")
+    _component = 1;
+  else if (_component_enum == "z")
+    _component = 2;
```
`getEnum()` seems to be converting into C++ enums, I am converting into `unsigned int`. I do not need `_component` to be C++ enum...

andrsd replied:
I guess, I could use `operator int()` instead...

permcody replied:
Yeah, I know we've created an X, Y, Z enum elsewhere and then used it for indexing (work for the old style enums). Maybe it's time to just create a Component enum in MOOSE and use that everywhere. This pattern appears a lot. Do what you think is best here.

andrsd replied:
> Maybe it's time to just create a Component enum in MOOSE and use that everywhere.

Yeah, sounds good to me.

> Do what you think is best here.

I will use with `operator int()`

andrsd committed:
Refactoring the setup of penetration locator

PenetrationAux, GapValueAux and NodaFaceConstraint had the same code to
setup the PenetrationLocator they used via their input parameters.  This
common code was moved inside PenetrationLocator.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/152395/) on b702b9d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10556/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10556/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1b1c33ba382fa0ee00ddbacc172144c1b360bbb4`

permcody committed:
Reworking test due to integration failure

refs #10668

andrsd commented:
I will have more doco in a moment...

permcody commented:
Just realized I screwed my commit. I can fixup mine (missing newline before EOF) before you push.

permcody reviewed (changes requested):
Per our discussion, we'd like to see the global all-to-all communication dropped in favor of using custom built ranges over non-local elements. When we get to DistributedMesh, we can add a GeometricRelationshipManager to the NodalNormals object to make sure we have the elements we need for this calculation. It'll be much more scalable this way.

andrsd commented:
Since the required changes are quite large and it will take me some time to finish, I am going to close this so it is not cluttering the PR list.
